src/mutate/drawDefinitions/positionGovernor/qualifierProgression.ts

```diff
@@ -0,0 +1,163 @@
+  const appliedPolicies =
+    getAppliedPolicies({
+      tournamentRecord,
+      drawDefinition,
+      structure,
+      event,
+    }).appliedPolicies ?? {};
+
+  const policy = appliedPolicies[POLICY_TYPE_POSITION_ACTIONS];
+  const requireCompletedStructures = policy?.requireCompletedStructures;
+
+  const { qualifierPositions, positionAssignments } = structureAssignedDrawPositions({ structure });
+
+  if (!qualifierPositions.length)
+    return decorateResult({ result: { error: NO_DRAW_POSITIONS_AVAILABLE_FOR_QUALIFIERS } });
+
+  const assignedParticipantIds = positionAssignments.map((assignment) => assignment.participantId).filter(Boolean);
+
+  const { relevantLinks: eliminationSourceLinks } =
+    getSourceStructureIdsAndRelevantLinks({
+      linkType: WINNER, // WINNER of qualifying structures will traverse link
+      drawDefinition,
+      structureId: structure.structureId,
+    }) || {};
+
+  const { relevantLinks: roundRobinSourceLinks } =
+    getSourceStructureIdsAndRelevantLinks({
+      linkType: POSITION, // link will define how many finishingPositions traverse the link
+      drawDefinition,
+      structureId: structure.structureId,
+    }) || {};
+
+  for (const sourceLink of eliminationSourceLinks) {
+    const structure = drawDefinition.structures?.find(
+      (structure) => structure.structureId === sourceLink.source.structureId,
+    );
+    if (structure?.stage !== QUALIFYING) continue;
+
+    const structureCompleted = isCompletedStructure({
+      structureId: sourceLink.source.structureId,
+      drawDefinition,
+    });
+
+    if (!requireCompletedStructures || structureCompleted) {
+      const qualifyingRoundNumber = structure.qualifyingRoundNumber;
+      const { matchUps } = getAllStructureMatchUps({
+        matchUpFilters: {
+          ...(qualifyingRoundNumber && {
+            roundNumbers: [qualifyingRoundNumber],
+          }),
+          hasWinningSide: true,
+        },
+        afterRecoveryTimes: false,
+        inContext: true,
+        structure,
+      });
+
+      for (const matchUp of matchUps) {
+        const winningSide = matchUp.sides.find((side) => side?.sideNumber === matchUp.winningSide);
+        const relevantSide = matchUp.matchUpStatus === BYE && matchUp.sides?.find(({ participantId }) => participantId);
+
+        if (winningSide || relevantSide) {
+          const { participantId } = winningSide || relevantSide || {};
+          if (participantId && !assignedParticipantIds.includes(participantId)) {
+            qualifyingParticipantIds.push(participantId);
+          }
+        }
+      }
+    }
+  }
+
+  for (const sourceLink of roundRobinSourceLinks) {
+    const structure = drawDefinition?.structures?.find(
+      (structure) => structure.structureId === sourceLink.source.structureId,
+    );
+    if (structure?.stage !== QUALIFYING) continue;
+
+    const structureCompleted = isCompletedStructure({
```

Comment:
For Round Robins structures must always be completed because otherwise there will be no finishingPosition available, so checking whether completion is required is irrelevant